### PR TITLE
review/jm-initial

### PR DESCRIPTION
**Pull request by fivetran-joemarkiewicz**
This PR serves as the initial review of the Twilio Source package. Please read through the comments and address them individually. Please do not resolve the comments as I will be sure to close them during my re-review. All changes should be made to the `main` branch. This branch is not intended to be merged.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,7 +10,7 @@
         <img src="https://img.shields.io/badge/Contributions-welcome-blueviolet" /></a>
 </p>
 
-# Twilio Source dbt Package ([docs](https://fivetran.github.io/dbt_twilio_source/))
+# Twilio Source dbt Package ([docs](https://fivetran.github.io/dbt_twilio_source/))
```
Comment (fivetran-joemarkiewicz) [inline README.md:13]: Remember to host the docs on GitHub pages

File: dbt_project.yml [hunk 2]
```diff
@@ -6,7 +6,7 @@ config-version: 2
 models:
   twilio_source:
     +materialized: table
-    +schema: stg_twilio
+    +schema: stg_twilio
```
Comment (fivetran-joemarkiewicz) [inline dbt_project.yml:9]: Our current standard is to name the staging model schema as `[platform]_source` to be more consistent with the name of the package. 

[See how this is done in Netsuite](https://github.com/fivetran/dbt_netsuite_source/blob/4311576a1a1fe7b8d750e8c9b8c41b7cbec73ad3/dbt_project.yml#L9).

File: macros/get_call_columns.sql [hunk 3]
```diff
@@ -26,17 +26,17 @@
...
+    {{ columns.append({"name": "FROM", "datatype": dbt.type_string(), "quote": True, "alias": "call_from"}) }} 
+{% else %} 
+    {{ columns.append({"name": "from", "datatype": dbt.type_string(), "quote": True, "alias": "call_from"}) }} 
+{% endif %}
+ 
+{% if target.type == 'snowflake' %} 
+    {{ columns.append({"name": "TO", "datatype": dbt.type_string(), "quote": True, "alias": "call_to"}) }} 
+{% else %} 
+    {{ columns.append({"name": "to", "datatype": dbt.type_string(), "quote": True, "alias": "call_to"}) }} 
+{% endif %}
```
Comment (fivetran-joemarkiewicz) [inline macros/get_call_columns.sql:39]: Is there any reason these are in two separate conditionals? I imagine we could make this dryer by combining them.

Comment (fivetran-joemarkiewicz) [inline macros/get_call_columns.sql:39]: The trailing comma could potentially cause issue. We should remove the comma from the last item in the macro.

File: macros/get_message_columns.sql [hunk 4]
```diff
@@ -19,17 +19,17 @@
...
-
-{% if target.type == 'snowflake' %}
-    {{ columns.append({"name": "TO", "datatype": dbt.type_string(), "quote": True, "alias": "message_to"}) }}
-{% else %}
-    {{ columns.append({"name": "to", "datatype": dbt.type_string(), "quote": True, "alias": "message_to"}) }}
-{% endif %},
+    {{ columns.append({"name": "TO", "datatype": dbt.type_string(), "quote": True, "alias": "message_to"}) }} 
+{% else %} 
+    {{ columns.append({"name": "to", "datatype": dbt.type_string(), "quote": True, "alias": "message_to"}) }} 
+{% endif %}
```
Comment (fivetran-joemarkiewicz) [inline macros/get_message_columns.sql:32]: Same comment from previous macro. Is there a reason we do not combine these conditionals? Also, be sure to remove the trailing comma.

File: models/src_twilio.yml [hunk 5]
```diff
@@ -8,9 +8,9 @@ sources:
     loader: fivetran
     loaded_at_field: _fivetran_synced
       
-    freshness:
-      warn_after: {count: 72, period: hour}
-      error_after: {count: 168, period: hour}
+    freshness: 
+      warn_after: {count: 72, period: hour} 
+      error_after: {count: 168, period: hour}
```
Comment (fivetran-joemarkiewicz) [inline models/src_twilio.yml:13]: Was there any specific reason these freshness tests were aligned on for **all** source tables? I imagine this may not be consistent with the freshness requirements of all tables. What are your thoughts?

File: models/src_twilio.yml [hunk 6]
```diff
@@ -43,8 +43,8 @@ sources:
             description: timestamp of when fivetran synced a record. 
 
           - name: account_id
-            description: >
-              The unique identifier of the account associated with the address.
+            description: > 
+              The unique identifier of the account associated with the address.
```
Comment (fivetran-joemarkiewicz) [inline models/src_twilio.yml:47]: Not a big issue, but is there a reason there is not consistency with the description format? Some use the `>` to put the description on a new line and others don't. Typically I have only seen us use the `>` if the description is exceptionally long. But these seem pretty short. Can we update the docs to be consistent with the format.

File: models/src_twilio.yml [hunk 7]
```diff
@@ -203,7 +203,7 @@ sources:
         
       - name: incoming_phone_number
         identifier: "{{ var('twilio_incoming_phone_number_identifier', 'incoming_phone_number' )}}"
-        description: Each record represents a Twilio phone number provisioned from Twilio, ported or hosted to Twili
+        description: Each record represents a Twilio phone number provisioned from Twilio, ported or hosted to Twilio
```
Comment (fivetran-joemarkiewicz) [inline models/src_twilio.yml:206]: Misspelling update for Twilio

File: models/src_twilio.yml [hunk 8]
```diff
@@ -437,7 +437,7 @@ sources:
             description: timestamp of when fivetran synced a record. 
 
           - name: created_at
-            description: 	The date that this resource was created, given in RFC 2822 format.
+            description: The date that this resource was created, given in RFC 2822 format.
```
Comment (fivetran-joemarkiewicz) [inline models/src_twilio.yml:440]: This was causing an error for me. This extra whitespace needed to be removed in order to run the package on my end.

File: models/src_twilio.yml [hunk 9]
```diff
@@ -456,7 +456,7 @@ sources:
         description: Each record provides details about resource consumption in Twilio.
         columns:
           - name: _fivetran_synced
-            description: timestamp of when fivetran synced a record. 
+            description: Timestamp of when fivetran synced a record.
```
Comment (fivetran-joemarkiewicz) [inline models/src_twilio.yml:459]: Grammar adjustment

File: models/stg_twilio.yml [hunk 10]
```diff
@@ -6,8 +6,8 @@ models:
     tests:
       - dbt_utils.unique_combination_of_columns:
           combination_of_columns:
-            - account_id
-            - updated_at
+            - account_id 
+            - updated_at
```
Comment (fivetran-joemarkiewicz) [inline models/stg_twilio.yml:10]: I agree that we should test on the uniqueness of these two fields. However, should we also test that they are not null? I see that we test not_null for other models that only have one field.

File: models/stg_twilio.yml [hunk 11]
```diff
@@ -285,9 +285,9 @@ models:
     tests:
       - dbt_utils.unique_combination_of_columns:
           combination_of_columns:
-            - account_id
-            - message_id
-            - messaging_service_id
+            - account_id 
+            - message_id 
+            - messaging_service_id
```
Comment (fivetran-joemarkiewicz) [inline models/stg_twilio.yml:290]: Same not null test comment

File: models/stg_twilio.yml [hunk 12]
```diff
@@ -349,8 +349,8 @@ models:
     tests: 
       - dbt_utils.unique_combination_of_columns:
           combination_of_columns:
-            - account_id
-            - messaging_service_id
+            - account_id 
+            - messaging_service_id
```
Comment (fivetran-joemarkiewicz) [inline models/stg_twilio.yml:353]: Same not null test comment

File: models/stg_twilio.yml [hunk 13]
```diff
@@ -447,7 +447,7 @@ models:
     description: A table with information on each single verified number that may be used as a caller ID when making outgoing calls via the REST API and within the TwiML <Dial> verb.
     columns:
       - name: _fivetran_synced
-        description: timestamp of when fivetran synced a record. 
+        description: Timestamp of when fivetran synced a record.
```
Comment (fivetran-joemarkiewicz) [inline models/stg_twilio.yml:450]: Grammar update

File: models/stg_twilio.yml [hunk 14]
```diff
@@ -456,7 +456,7 @@ models:
           - not_null
         
       - name: created_at
-        description: 	The date that this resource was created, given in RFC 2822 format.
+        description: The date that this resource was created, given in RFC 2822 format.
```
Comment (fivetran-joemarkiewicz) [inline models/stg_twilio.yml:459]: Similarly, I needed to remove the whitespace in order for the package to run on my end.

File: models/stg_twilio.yml [hunk 15]
```diff
@@ -473,10 +473,10 @@ models:
...
       - dbt_utils.unique_combination_of_columns:
           combination_of_columns:
-            - category
-            - start_date
-            - end_date
-            - account_id
+            - category 
+            - start_date 
+            - end_date 
+            - account_id
```
Comment (fivetran-joemarkiewicz) [inline models/stg_twilio.yml:479]: Same not null test comment

File: models/stg_twilio__account_history.sql [hunk 16]
```diff
@@ -1,5 +1,5 @@
---To disable this model, set the using_domain_names variable within your dbt_project.yml file to False.
-{{ config(enabled=var('using_twilio_account_history', True)) }}
+--To disable this model, set the using_domain_names variable within your dbt_project.yml file to False. 
+{{ config(enabled=var('using_twilio_account_history', True)) }}
```
Comment (fivetran-joemarkiewicz) [inline models/stg_twilio__account_history.sql:2]: Is there a reason we are allowing customers to disable this model? I imagine this is a pretty necessary model for the downstream transformations.

Comment (fivetran-joemarkiewicz) [reply models/stg_twilio__account_history.sql:2]: Actually I noticed you added the disable config for **all** the models. Is there any reason in particular this was decided?

Additionally, I noticed the enabled configs were not added to the tmp models. If we do decide to include them, we should be sure to add the config to the tmp models. Otherwise the disabling will not function as expected.

File: README.md [hunk 17]
```diff
@@ -26,7 +26,7 @@
 To use this dbt package, you must have the following:
 
 - At least one Fivetran Twilio connector syncing data into your destination.
-- A **BigQuery**, **Snowflake**, **Redshift**, or **PostgreSQL** destination.
+- A **BigQuery**, **Snowflake**, **Redshift**, or **PostgreSQL** destination.
```
Comment (fivetran-joemarkiewicz) [inline README.md:29]: It looks like Databricks is not included. I also noticed Databricks is not included in the integration tests. We should make all new packages compatible with Databricks. Please update the package to be compatible and test Databricks.

File: README.md [hunk 18]
```diff
@@ -47,7 +47,7 @@ vars:
   twilio_schema: your_schema_name 
 ```
 
-## Step 4: Enabling/Disabling Models
+## Step 4: Enabling/Disabling Models
```
Comment (fivetran-joemarkiewicz) [inline README.md:50]: Similar comment asking why we are allowing customers to disable every model in the package?

File: README.md [hunk 19]
```diff
@@ -71,7 +71,7 @@ vars:
 
 ### Changing the Build Schema
 
-By default, this package will build the Twilio staging models within a schema titled (<target_schema> + `_stg_twilio`) in your target database. If this is not where your would like you Twilio staging data to be written to, add the following configuration to your `dbt_project.yml` file:
+By default, this package will build the Twilio staging models within a schema titled (<target_schema> + `_stg_twilio`) in your target database. If this is not where your would like you Twilio staging data to be written to, add the following configuration to your `dbt_project.yml` file:
```
Comment (fivetran-joemarkiewicz) [inline README.md:74]: Be sure to update this section after updating the custom schema config.

File: README.md [hunk 20]
```diff
@@ -96,7 +96,7 @@ vars:
 Fivetran offers the ability for you to orchestrate your dbt project through [Fivetran Transformations for dbt Core™](https://fivetran.com/docs/transformations/dbt). Learn how to set up your project for orchestration through Fivetran in our [Transformations for dbt Core setup guides](https://fivetran.com/docs/transformations/dbt#setupguide).
 </details>
 
-# 🔍 Does this package have dependencies?
+# 🔍 Does this package have dependencies?
```
Comment (fivetran-joemarkiewicz) [inline README.md:99]: Reminder to add spark-utils if it is needed after making the package databricks compatible.

File: README.md [hunk 21]
```diff
@@ -121,4 +121,4 @@ We highly encourage and welcome contributions to this package. Check out [this d
 # 🏪 Are there any resources available?
 - If you have questions or want to reach out for help, please refer to the [GitHub Issue](https://github.com/fivetran/dbt_twilio_source/issues/new/choose) section to find the right avenue of support for you.
 - If you would like to provide feedback to the dbt package team at Fivetran or would like to request a new dbt package, fill out our [Feedback Form](https://www.surveymonkey.com/r/DQ7K7WW).
-- Have questions or want to just say hi? Book a time during our office hours [on Calendly](https://calendly.com/fivetran-solutions-team/fivetran-solutions-team-office-hours) or email us at solutions@fivetran.com.
\ No newline at end of file
+- Have questions or want to just say hi? Book a time during our office hours [on Calendly](https://calendly.com/fivetran-solutions-team/fivetran-solutions-team-office-hours) or email us at solutions@fivetran.com.
```
Comment (fivetran-joemarkiewicz) [inline README.md:124]: We are trying to direct more folks to the Fivetran Community (dbt User Group) rather than directly to the office hours. 

Would you be able to update this section to instead look like the following (from our Zendesk README)
image

File: .github/ISSUE_TEMPLATE/config.yml [hunk 22]
```diff
@@ -1,7 +1,7 @@
 contact_links:
-  - name: Ask a question during our office hours
-    url: https://calendly.com/fivetran-solutions-team/fivetran-solutions-team-office-hours
-    about: Schedule time during the external office hours block with the Fivetran Analytics Engineering team for support
+  - name: Ask a question during our office hours 
+    url: https://calendly.com/fivetran-solutions-team/fivetran-solutions-team-office-hours 
+    about: Schedule time during the external office hours block with the Fivetran Analytics Engineering team for support
```
Comment (fivetran-joemarkiewicz) [inline .github/ISSUE_TEMPLATE/config.yml:4]: Similar to the comment about removing the direct link to the office hours, can we update this section to remove the office hours link and resemble the [format we have in Zendesk](https://github.com/fivetran/dbt_zendesk/blob/main/.github/ISSUE_TEMPLATE/config.yml) to direct users to the community

**Review by fivetran-joemarkiewicz — COMMENTED** [review]
@fivetran-reneeli I just did an initial review and have a few comments. Please review and respond to all comments. Let me know if you have any questions and when this is ready for re-review.

File: models/stg_twilio__account_history.sql [hunk 23]
```diff
@@ -29,7 +29,7 @@ final as (
         owner_account_id,
         status,
         type,
-        updated_at
+        updated_at
```
Comment (fivetran-joemarkiewicz) [inline models/stg_twilio__account_history.sql:32]: Since this is a history table, should we create a most recent field? Similar to what we have done in packages in the past?

Additionally, if we do take this approach then we should filter out old records

File: dbt_project.yml (hunk 2 above)
Comment (fivetran-reneeli) [reply dbt_project.yml:9]: updated

File: macros/get_call_columns.sql (hunk 3 above)
Comment (fivetran-reneeli) [reply macros/get_call_columns.sql:39]: good point- updated

Comment (fivetran-reneeli) [reply macros/get_call_columns.sql:39]: removed

File: macros/get_message_columns.sql (hunk 4 above)
Comment (fivetran-reneeli) [reply macros/get_message_columns.sql:32]: done!

File: models/src_twilio.yml (hunk 6 above)
Comment (fivetran-reneeli) [reply models/src_twilio.yml:47]: Yeah, when I used AI to generate the descriptions, it automatically added them, didn't think it was worth the time to remove it but I did now!

File: models/src_twilio.yml (hunk 7 above)
Comment (fivetran-reneeli) [reply models/src_twilio.yml:206]: fixed

File: models/src_twilio.yml (hunk 8 above)
Comment (fivetran-reneeli) [reply models/src_twilio.yml:440]: Extra whitespace removed

File: models/src_twilio.yml (hunk 9 above)
Comment (fivetran-reneeli) [reply models/src_twilio.yml:459]: updated all of these instances

File: models/stg_twilio.yml (hunk 10 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:10]: that is true, added!

File: models/stg_twilio.yml (hunk 11 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:290]: added not null tests to each of these fields, and unique test to message_id

File: models/stg_twilio.yml (hunk 12 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:353]: added not null and unique test to messaging_service_id , and not null test to account_id for `stg_twilio__messaging_service`

File: models/stg_twilio.yml (hunk 13 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:450]: updated

File: models/stg_twilio.yml (hunk 14 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:459]: fixed

File: models/stg_twilio.yml (hunk 15 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:479]: added not null tests to each of these fields

File: models/stg_twilio__account_history.sql (hunk 16 above)
Comment (fivetran-reneeli) [reply models/stg_twilio__account_history.sql:2]: I added them in case not every customer uses that table. I suppose for account_history, that's pretty much a given. It's more an issue for call or message related table, if the customer doesn't use that certain service. I'll removed the disable config for account history and the others that will be massively used, will think through this a bit more

File: README.md (hunk 19 above)
Comment (fivetran-reneeli) [reply README.md:74]: udpated

File: .github/ISSUE_TEMPLATE/config.yml (hunk 22 above)
Comment (fivetran-reneeli) [reply .github/ISSUE_TEMPLATE/config.yml:4]: updated

File: models/stg_twilio__account_history.sql (hunk 23 above)
Comment (fivetran-reneeli) [reply models/stg_twilio__account_history.sql:32]: good point -- just added the filter to select for most recent

File: README.md (hunk 21 above)
Comment (fivetran-reneeli) [reply README.md:124]: updated

File: README.md (hunk 20 above)
Comment (fivetran-reneeli) [reply README.md:99]: added

File: models/src_twilio.yml (hunk 5 above)
Comment (fivetran-reneeli) [reply models/src_twilio.yml:13]: Good question. With this I just referenced other packages like klaviyo and iterable. But I suppose it doesn't make sense for example for account_history to use this frequent of a freshness test. I therefore moved it to call, message, and usage_record and changed it to a 24-hour warning, based on what I see in the sample customer data

Comment (fivetran-reneeli) [reply models/src_twilio.yml:13]: Let me know what you think!

File: README.md (hunk 18 above)
Comment (fivetran-reneeli) [reply README.md:50]: addressed above

File: models/stg_twilio.yml (hunk 11 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:290]: update - removed the not null test from `messaging_service_id`. Checking the data, looks like you _don't_ have to be using that field, so it _can_ be null

File: README.md (hunk 17 above)
Comment (fivetran-reneeli) [reply README.md:29]: add databricks

File: models/stg_twilio__account_history.sql (hunk 16 above)
Comment (fivetran-reneeli) [reply models/stg_twilio__account_history.sql:2]: added disable to the necessary tmp files

File: README.md (hunk 1 above)
Comment (fivetran-reneeli) [reply README.md:13]: updaetd

File: models/stg_twilio.yml (hunk 10 above)
Comment (fivetran-joemarkiewicz) [reply models/stg_twilio.yml:10]: It looks like a not null was only added for the account_id, should you also add it for updated_at?

File: models/stg_twilio.yml (hunk 11 above)
Comment (fivetran-joemarkiewicz) [reply models/stg_twilio.yml:290]: Good call! Is there a reason though `message_id` has it's own unique test as well? I feel that defeats the purpose of the unique combo test.

File: models/stg_twilio.yml (hunk 12 above)
Comment (fivetran-joemarkiewicz) [reply models/stg_twilio.yml:353]: Why is the unique test being added to messaging_service_id? If that is the case, why are we also doing a unique combo of columns test?

File: models/stg_twilio__account_history.sql (hunk 16 above)
Comment (fivetran-joemarkiewicz) [reply models/stg_twilio__account_history.sql:2]: It doesn't look like the disable configs were added to the tmp files.

File: README.md (hunk 20 above)
Comment (fivetran-joemarkiewicz) [reply README.md:99]: It actually looks like you didn't include spark-utils in the packages.yml. If spark-utils is not needed then we won't need to include it in the dep list.

That being said, you likely will need it in the transform package so it may be worthwhile to also include here. Regardless, be sure to add spark-utils to the packages.yml.

File: models/stg_twilio__account_history.sql (hunk 23 above)
Comment (fivetran-joemarkiewicz) [reply models/stg_twilio__account_history.sql:32]: Actually now that I think about this, we usually filter out the historical records when we bring the table in in the transform package. The most recent record field is great, but can we remove the filter. Apologies for the flip flop here.

File: models/src_twilio.yml (hunk 5 above)
Comment (fivetran-joemarkiewicz) [reply models/src_twilio.yml:13]: I think this is a good approach. However, can we maybe adjust the 24 hour warn window to 72 instead. Just to give a little bit more buffer.

Finally, be sure to add the [disable config for sources](https://github.com/fivetran/dbt_zendesk_source/blob/8b78b0d3a02d168abbf1afefed4cf7e114dad313/models/src_zendesk.yml#L94-L95) that can be disabled (like calls in this case)

**Comment by fivetran-joemarkiewicz** [comment]
@fivetran-reneeli thanks for replying to my initial comments. I just went through and replied to your comments. There are some where additional changes need to be made. Let me know once those are updated. Thanks.

File: macros/get_call_columns.sql (hunk 3 above)
Comment (fivetran-joemarkiewicz) [reply macros/get_call_columns.sql:39]: Thanks for applying this update. However, can we make sure the trailing commas are appropriate. It looks like the conditional blocks don't have the commas. It should look like the following.

```sql
    {"name": "trunk_id", "datatype": dbt.type_string()},
    {"name": "updated_at", "datatype": dbt.type_timestamp()},
] %}

{% if target.type == 'snowflake' %}
    {{ columns.append({"name": "FROM", "datatype": dbt.type_string(), "quote": True, "alias": "call_from"}) }},
    {{ columns.append({"name": "TO", "datatype": dbt.type_string(), "quote": True, "alias": "call_to"}) }}
    {% else %}
    {{ columns.append({"name": "from", "datatype": dbt.type_string(), "quote": True, "alias": "call_from"}) }},
    {{ columns.append({"name": "to", "datatype": dbt.type_string(), "quote": True, "alias": "call_to"}) }}
{% endif %}
```

File: macros/get_message_columns.sql (hunk 4 above)
Comment (fivetran-joemarkiewicz) [reply macros/get_message_columns.sql:32]: Same comment as the previous

File: macros/get_call_columns.sql (hunk 3 above)
Comment (fivetran-reneeli) [reply macros/get_call_columns.sql:39]: Yeah I thought about that! But I noticed in our other packages (ex Amplitude, Jira, Stripe) [Jira ex](https://github.com/fivetran/dbt_jira_source/blob/main/macros/get_issue_field_history_columns.sql), we don't do that. I'll make the change since it doesn't hurt, but we should keep it mind for our other packages..

File: models/src_twilio.yml (hunk 5 above)
Comment (fivetran-reneeli) [reply models/src_twilio.yml:13]: done and done!

File: models/stg_twilio.yml (hunk 10 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:10]: added for updated_at

File: models/stg_twilio.yml (hunk 11 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:290]: Oh that's true, I could see a chance for 2 of the same message ids across different accounts. Good catch. Removed

File: models/stg_twilio.yml (hunk 12 above)
Comment (fivetran-reneeli) [reply models/stg_twilio.yml:353]: True, same logic applies as for `stg_twilio__message`. Removed the unique test for just messaging_service_id

File: models/stg_twilio__account_history.sql (hunk 16 above)
Comment (fivetran-reneeli) [reply models/stg_twilio__account_history.sql:2]: oops- added now

File: README.md (hunk 20 above)
Comment (fivetran-reneeli) [reply README.md:99]: thanks for catching this, added to packages.yml

File: models/stg_twilio__account_history.sql (hunk 23 above)
Comment (fivetran-reneeli) [reply models/stg_twilio__account_history.sql:32]: I see, I'll keep the `is_most_recent_record` but move to using as a filter in transforms

**Comment by fivetran-joemarkiewicz** [comment]
Thanks for applying the updates @fivetran-reneeli! This looks great! Closing out the PR
